components/accelerator/nvidia/error/xid/component.go

```diff
@@ -5,150 +5,240 @@ package xid
-	if err != nil {
-		return nil, err
-	}
+func (c *XIDComponent) Name() string { return nvidia_component_error_xid_id.Name }
 
-	var dmesgComponent *dmesg.Component
-	if o, ok := dmesgC.(interface{ Unwrap() interface{} }); ok {
-		if unwrapped, ok := o.Unwrap().(*dmesg.Component); ok {
-			dmesgComponent = unwrapped
-		} else {
-			return nil, fmt.Errorf("expected *dmesg.Component, got %T", dmesgC)
-		}
-	}
-	dmesgTailResults, err := dmesgComponent.TailScan()
+func (c *XIDComponent) Start() error {
+	watcher, err := pkg_dmesg.NewWatcher()
 	if err != nil {
-		return nil, err
+		log.Logger.Errorw("failed to create dmesg watcher", "error", err)
+		return nil
 	}
-
-	o := &Output{}
-	for _, logItem := range dmesgTailResults.TailScanMatched {
-		if logItem.Error != nil {
+	for {
+		backoffTime := 1 * time.Second
+		osComponent, err := components.GetComponent(os_id.Name)
+		if err != nil {
+			log.Logger.Errorw("failed to get os component", "error", err)
+			time.Sleep(backoffTime)
 			continue
 		}
-		if logItem.Matched == nil {
+		osEvents, err := osComponent.Events(c.rootCtx, time.Now().Add(-3*24*time.Hour))
+		if err != nil {
+			log.Logger.Errorw("failed to get os states", "error", err)
+			time.Sleep(backoffTime)
 			continue
 		}
-		if logItem.Matched.Name != dmesg.EventNvidiaNVRMXid {
+		if len(osEvents) < 1 {
+			log.Logger.Debugw("no os states found")
+			time.Sleep(backoffTime)
 			continue
 		}
-
-		ev, err := nvidia_query_xid.ParseDmesgLogLine(logItem.Time, logItem.Line)
+		localEvents, err := c.store.Get(c.rootCtx, time.Time{})
 		if err != nil {
-			return nil, err
+			log.Logger.Errorw("failed to get all events", "error", err)
+			time.Sleep(backoffTime)
+			continue
 		}
-		o.DmesgErrors = append(o.DmesgErrors, ev)
-	}
-
-	last, err := c.poller.LastSuccess()
-
-	// no data yet from realtime xid poller
-	// just return whatever we got from dmesg
-	if err == query.ErrNoData {
-		log.Logger.Debugw("nothing found in last state (no data collected yet)", "component", nvidia_component_error_xid_id.Name)
-		return o, nil
+		events := mergeEvents(osEvents, localEvents)
+		c.Lock()
+		c.currState = EvolveHealthyState(events)
+		c.Unlock()
+		break
 	}
-
-	// something went wrong in the poller
-	// just return an error to surface the issue
-	if err != nil {
-		return nil, err
-	}
-
-	ev, ok := last.Output.(*nvidia_query_nvml.XidEvent)
-	if !ok { // shoild never happen
-		return nil, fmt.Errorf("invalid output type: %T, expected nvidia_query_nvml.XidEvent", last.Output)
-	}
-	if ev != nil && ev.Xid > 0 {
-		o.NVMLXidEvent = ev
-
-		if lerr := c.poller.LastError(); lerr != nil {
-			log.Logger.Warnw("last query failed -- returning cached, possibly stale data", "error", lerr)
+	go func() {
+		ticker := time.NewTicker(DefaultStateUpdatePeriod)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-c.rootCtx.Done():
+				return
+			case <-ticker.C:
+				osComponent, err := components.GetComponent(os_id.Name)
```

Comment:
I'm curious because I remember we discussed whether we needed to pull the OS state regularly. At that time, we decided it wasn't necessary, but now it seems we need to pull the latest OS state after all.

Reply:
yes, during testing, figured that we can manually generate reboot event (without actually rebooting) and this will be skipped -> incorrect state. Even this is not likely to happen on prod, i think its better for Xid component to cover it..
And since we need to periodically update state anyway (to reconcile when events are purged), thought might as well just make the stateUpdate logic generic to both initialization and periodic invoking..Since this only introduce a minute level extra sqlite query, not too heavy..